lib/gears/shape.lua

```diff
@@ -41,14 +41,128 @@
 -- @themelib gears.shape
 ---------------------------------------------------------------------------
 local g_matrix = require( "gears.matrix" )
+local g_math   = require( "gears.math" )
 local unpack   = unpack or table.unpack -- luacheck: globals unpack (compatibility with Lua 5.1)
 local atan2    = math.atan2 or math.atan -- lua 5.3 compat
 local min      = math.min
 local cos      = math.cos
 local sin      = math.sin
+local pow      = math.pow -- luacheck: globals math.pow (math.pow is deprecated in lua5.3, but is still there)
+local abs      = math.abs
 
 local module = {}
 
+--- Add a squircle shape with only some of the corner are "circled" to the current path
+-- the squircle is not exactly as the definition
+-- it will expand to the shape's width and height
+-- kinda like an ellipse
+--
+-- @DOC_gears_shape_partial_squircle_EXAMPLE@
+--
+-- @param cr A cairo context
+-- @tparam number width The shape width
+-- @tparam number height The height constraint
+-- @tparam boolean tl If the top left corner is rounded
+-- @tparam boolean tr If the top right corner is rounded
+-- @tparam boolean br If the bottom right corner is rounded
+-- @tparam boolean bl If the bottom left corner is rounded
+-- @tparam number rate The "squareness" of the squircle
+-- @tparam number delta The "smoothness" of the shape
+-- @staticfct gears.shape.partial_squircle
+function module.partial_squircle(cr, width, height, tl, tr, br, bl, rate, delta)
+   -- rate ~ 2 can be used by icon
+   -- this shape doesn't really fit clients
+   -- but you can still use with rate ~ 32
+   rate = rate or 2
+   -- smaller the delta the smoother the shape
+   -- but probably more laggy
+   delta = delta or 0.05
```

Comment:
I thought that this shape will do a lot of `cr:line_to()` when being drawn so small `delta` might affect the performance.

Reply:
My suggestion is something like `delta = delta or math.max(width, height) / 20` where `20` is to be decided in I-do-not-know-how. I *feel*  like `0.05` might be a good number when the shape is used at a small size, but could lead to artifacts if I use it at full-screen-size. (Or if there are no artifacts: Then this does way too many steps at small sizes.)

Or am I misunderstanding something here and this suggestions makes no sense at all?

Reply:
I think my worry might be invalid here. I tried with `delta = 0.0005` (which is equivalent to `max(width, height) ~ 1920` and set a timer to send notification then immediately draw a wibox on the screen, doesn't took too long.

But it might be fast because I'm using luajit and a not too old laptop. sadly I don't have an old device to test it out.

Reply:
I did some quick benchmark: https://gist.github.com/psychon/a72b09cf94043a01621470b0f8a097b0

Result is:
```
Lua 5.3
0.5	18
0.05	130
0.005	1262
0.0005	9447
5e-05	89307
        squircle 0.5: 5.95582e-05 sec/iter (17669 iters, 3.644 sec for benchmark)
       squircle 0.05: 0.000364222 sec/iter (2858 iters, 1.041 sec for benchmark)
      squircle 0.005: 0.003471   sec/iter (306 iters, 1.065 sec for benchmark)
     squircle 0.0005: 0.0338544  sec/iter ( 30 iters, 1.049 sec for benchmark)
      squircle 5e-05: 0.336964   sec/iter (  3 iters, 1.353 sec for benchmark)
```
So... the number of points seems to increase by less than 10 if I divide the `rate` by 10. That's good, but also confuses me. Not what I had expected.

And for the times: Well, dividing the `rate` roughly increases the run time by factor 10. Dunno what this means for us.

Reply:
Which sizes would you consider small? `max(width, height) == 24` would still create more step than `delta = 0.05`

About the artifact, seems like the shape is going to be a bit rough when using `delta = 0.05` (please zoom at the bottom right border of the terminal to see it)
![image](https://user-images.githubusercontent.com/33273948/80813870-12fc6280-8bf5-11ea-8dae-71f34266d75b.png)

And this one is `delta = 1 / max(width, height)`, got little improvement on big shapes, but it does look smoother on small shapes. 
![image](https://user-images.githubusercontent.com/33273948/80814329-e4cb5280-8bf5-11ea-91d9-2a897dc437fe.png)

`delta = max(width, height) / 20` will create weird shape when your shape's size is bigger than 20. 

> So... the number of points seems to increase by less than 10 if I divide the rate by 10. That's good, but also confuses me. Not what I had expected.

We are approximating `pi` (this shape was draw using polar cordintae), I think there would be error for that

> And for the times: Well, dividing the rate roughly increases the run time by factor 10. Dunno what this means for us.

My suggestion is: use `max(width, height)`  and have a constraint on delta (if `delta` is less then `something` then throw an error or just use `delta = something`). What do you think?

Reply:
> delta = max(width, height) / 20 will create weird shape when your shape's size is bigger than 20.

Yeah, sorry for that suggestion. I got my math backwards here.

> What do you think?

Pick whatever you feel is best. I was just leaving some drive-by comments about things that made me wonder.

Reply:
I think I'll go with this:

```lua
delta = delta or 1 / max(width, height)
delta = delta > 0.01 and delta or 0.01
```

Since this shape will mostly be used by icons, small size would be expected so the 0.01 limit wouldn't affect much.